File: api/routes/reports.py
"""
/**
 * 
 *  ┌─────────────────────────────────────┐
 *  │         REPORTS API ROUTES          │
 *  └─────────────────────────────────────┘
 *  API endpoints for managing AI analysis reports
 * 
 *  Provides RESTful endpoints for creating, reading, updating,
 *  and deleting AI analysis reports.
 * 
 *  Parameters:
 *  - None
 * 
 *  Returns:
 *  - FastAPI router instance
 * 
 *  Notes:
 *  - Follows REST conventions
 *  - Returns JSON responses
 *  - Handles errors gracefully
 */
"""

from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

from core.models import ReportModel, AIAction, AIAnalysisStatus
from data.repositories.reports import get_reports_repository
from debugger import debug_info, debug_error, debug_success


# Pydantic models for request/response
class ReportCreate(BaseModel):
    """Request model for creating a report"""
    symbol: str
    ai_summary: str = Field(alias="AISummary")
    ai_action: str = Field(alias="AIAction")
    ai_confidence: float = Field(alias="AIConfidence")
    ai_event_time: Optional[str] = Field(None, alias="AIEventTime")
    ai_levels: Optional[str] = Field(None, alias="AILevels")
    ai_analysis_status: Optional[str] = Field("completed", alias="AIAnalysisStatus")
    
    class Config:
        populate_by_name = True


class ReportUpdate(BaseModel):
    """Request model for updating a report"""
    ai_summary: Optional[str] = Field(None, alias="AISummary")
    ai_action: Optional[str] = Field(None, alias="AIAction")
    ai_confidence: Optional[float] = Field(None, alias="AIConfidence")
    ai_event_time: Optional[str] = Field(None, alias="AIEventTime")
    ai_levels: Optional[str] = Field(None, alias="AILevels")
    ai_analysis_status: Optional[str] = Field(None, alias="AIAnalysisStatus")
    
    class Config:
        populate_by_name = True


class ReportResponse(BaseModel):
    """Response model for report data"""
    id: int
    timeFetched: str
    symbol: str
    AISummary: str
    AIAction: str
    AIConfidence: float
    AIEventTime: Optional[str]
    AILevels: Optional[str]
    AIAnalysisStatus: str


# Create router
router = APIRouter(prefix="/api/reports", tags=["reports"])


@router.post("/", response_model=ReportResponse)
async def create_report(report_data: ReportCreate):
    """
     ┌─────────────────────────────────────┐
     │         CREATE_REPORT               │
     └─────────────────────────────────────┘
     Create a new AI analysis report
     
     Creates a report with the provided analysis data.
    """
    try:
        # Create report model
        report = ReportModel(
            time_fetched=datetime.now(),
            symbol=report_data.symbol.upper(),
            ai_summary=report_data.ai_summary,
            ai_action=AIAction(report_data.ai_action.upper()),
            ai_confidence=report_data.ai_confidence,
            ai_event_time=report_data.ai_event_time,
            ai_levels=report_data.ai_levels,
            ai_analysis_status=AIAnalysisStatus(report_data.ai_analysis_status)
        )
        
        # Save to database
        repo = get_reports_repository()
        report_id = repo.create(report)
        
        # Get created report
        created_report = repo.get_by_id(report_id)
        if not created_report:
            raise HTTPException(status_code=500, detail="Failed to retrieve created report")
        
        debug_success(f"Created report {report_id} for {report.symbol}")
        return ReportResponse(**created_report.to_dict())
        
    except ValueError as e:
        debug_error(f"Invalid report data: {e}")
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        debug_error(f"Error creating report: {e}")
        raise HTTPException(status_code=500, detail="Failed to create report")


@router.get("/{report_id}", response_model=ReportResponse)
async def get_report(report_id: int):
    """
     ┌─────────────────────────────────────┐
     │           GET_REPORT                │
     └─────────────────────────────────────┘
     Get a specific report by ID
    """
    repo = get_reports_repository()
    report = repo.get_by_id(report_id)
    
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    return ReportResponse(**report.to_dict())


@router.get("/", response_model=List[ReportResponse])
async def get_reports(
    symbol: Optional[str] = Query(None, description="Filter by symbol"),
    limit: Optional[int] = Query(100, description="Maximum number of reports"),
    offset: int = Query(0, description="Number of reports to skip")
):
    """
     ┌─────────────────────────────────────┐
     │          GET_REPORTS                │
     └─────────────────────────────────────┘
     Get reports with optional filtering
     
     Query parameters:
     - symbol: Filter by trading symbol
     - limit: Maximum number of results
     - offset: Pagination offset
    """
    repo = get_reports_repository()
    
    if symbol:
        reports = repo.get_by_symbol(symbol.upper(), limit=limit)
    else:
        reports = repo.get_all(limit=limit, offset=offset)
    
    return [ReportResponse(**report.to_dict()) for report in reports]


@router.get("/symbol/{symbol}", response_model=List[ReportResponse])
async def get_reports_by_symbol(
    symbol: str,
    limit: Optional[int] = Query(10, description="Maximum number of reports")
):
    """
     ┌─────────────────────────────────────┐
     │      GET_REPORTS_BY_SYMBOL          │
     └─────────────────────────────────────┘
     Get all reports for a specific symbol
    """
    repo = get_reports_repository()
    reports = repo.get_by_symbol(symbol.upper(), limit=limit)
    
    if not reports:
        return []  # Return empty list instead of 404
    
    return [ReportResponse(**report.to_dict()) for report in reports]


@router.get("/symbol/{symbol}/latest", response_model=Optional[ReportResponse])
async def get_latest_report_by_symbol(symbol: str):
    """
     ┌─────────────────────────────────────┐
     │    GET_LATEST_REPORT_BY_SYMBOL      │
     └─────────────────────────────────────┘
     Get the most recent report for a symbol
    """
    repo = get_reports_repository()
    report = repo.get_latest_by_symbol(symbol.upper())
    
    if not report:
        return None  # Return null instead of 404
    
    return ReportResponse(**report.to_dict())


@router.put("/{report_id}", response_model=ReportResponse)
async def update_report(report_id: int, update_data: ReportUpdate):
    """
     ┌─────────────────────────────────────┐
     │         UPDATE_REPORT               │
     └─────────────────────────────────────┘
     Update an existing report
    """
    repo = get_reports_repository()
    
    # Check if report exists
    report = repo.get_by_id(report_id)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    # Build update dict
    updates = {}
    if update_data.ai_summary is not None:
        updates["AISummary"] = update_data.ai_summary
    if update_data.ai_action is not None:
        updates["AIAction"] = update_data.ai_action.upper()
    if update_data.ai_confidence is not None:
        updates["AIConfidence"] = update_data.ai_confidence
    if update_data.ai_event_time is not None:
        updates["AIEventTime"] = update_data.ai_event_time
    if update_data.ai_levels is not None:
        updates["AILevels"] = update_data.ai_levels
    if update_data.ai_analysis_status is not None:
        updates["AIAnalysisStatus"] = update_data.ai_analysis_status
    
    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")
    
    # Update report
    success = repo.update(report_id, updates)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to update report")
    
    # Get updated report
    updated_report = repo.get_by_id(report_id)
    debug_info(f"Updated report {report_id}")
    
    return ReportResponse(**updated_report.to_dict())


@router.delete("/{report_id}")
async def delete_report(report_id: int):
    """
     ┌─────────────────────────────────────┐
     │         DELETE_REPORT               │
     └─────────────────────────────────────┘
     Delete a report
    """
    repo = get_reports_repository()
    
    # Check if report exists
    report = repo.get_by_id(report_id)
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    # Delete report
    success = repo.delete(report_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to delete report")
    
    debug_info(f"Deleted report {report_id}")
    return {"message": "Report deleted successfully", "id": report_id}


@router.get("/high-confidence/{min_confidence}", response_model=List[ReportResponse])
async def get_high_confidence_reports(
    min_confidence: float = 0.8,
    limit: Optional[int] = Query(50, description="Maximum number of reports")
):
    """
     ┌─────────────────────────────────────┐
     │   GET_HIGH_CONFIDENCE_REPORTS       │
     └─────────────────────────────────────┘
     Get reports with high confidence scores
    """
    repo = get_reports_repository()
    reports = repo.get_high_confidence_reports(min_confidence)
    
    # Apply limit
    if limit:
        reports = reports[:limit]
    
    return [ReportResponse(**report.to_dict()) for report in reports]


@router.get("/summary/by-action", response_model=Dict[str, int])
async def get_reports_summary_by_action():
    """
     ┌─────────────────────────────────────┐
     │    GET_REPORTS_SUMMARY_BY_ACTION    │
     └─────────────────────────────────────┘
     Get count of reports grouped by action
    """
    repo = get_reports_repository()
    summary = repo.get_summary_by_action()
    return summary


@router.delete("/cleanup/{days}")
async def cleanup_old_reports(days: int = 30):
    """
     ┌─────────────────────────────────────┐
     │       CLEANUP_OLD_REPORTS           │
     └─────────────────────────────────────┘
     Delete reports older than specified days
    """
    if days < 1:
        raise HTTPException(status_code=400, detail="Days must be greater than 0")
    
    repo = get_reports_repository()
    deleted_count = repo.delete_old_reports(days)
    
    debug_info(f"Cleaned up {deleted_count} reports older than {days} days")
    return {
        "message": f"Deleted {deleted_count} old reports",
        "days": days,
        "count": deleted_count
    }

File: static/js/services/reports.js
/**
 * 
 *  ┌─────────────────────────────────────┐
 *  │         REPORTS SERVICE             │
 *  └─────────────────────────────────────┘
 *  Service for managing AI analysis reports
 * 
 *  Handles all client-side operations for reports including
 *  creating, fetching, updating, and deleting reports.
 * 
 *  Parameters:
 *  - None
 * 
 *  Returns:
 *  - ReportsService singleton instance
 * 
 *  Notes:
 *  - Uses apiClient for backend communication
 *  - Handles error states gracefully
 *  - Caches latest report per symbol
 */

import { apiClient } from '../core/api.js';

class ReportsService {
    constructor() {
        this.cache = new Map(); // Cache latest report per symbol
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │         CREATE REPORT               │
     *  └─────────────────────────────────────┘
     *  Creates a new AI analysis report
     * 
     *  Parameters:
     *  - reportData: Object with report fields
     * 
     *  Returns:
     *  - Created report object
     */
    async createReport(reportData) {
        try {
            const response = await apiClient.post('/api/reports/', reportData);
            
            // Update cache with new report
            if (response.symbol) {
                this.cache.set(response.symbol, response);
            }
            
            Debugger.success(`Report created for ${response.symbol}`);
            return response;
        } catch (error) {
            Debugger.error(`Failed to create report: ${error.message}`);
            throw error;
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │       GET REPORT BY ID              │
     *  └─────────────────────────────────────┘
     *  Fetches a specific report by ID
     * 
     *  Parameters:
     *  - reportId: Report ID
     * 
     *  Returns:
     *  - Report object or null
     */
    async getReportById(reportId) {
        try {
            const response = await apiClient.get(`/api/reports/${reportId}`);
            return response;
        } catch (error) {
            if (error.status === 404) {
                return null;
            }
            Debugger.error(`Failed to fetch report: ${error.message}`);
            throw error;
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │     GET REPORTS BY SYMBOL           │
     *  └─────────────────────────────────────┘
     *  Fetches all reports for a symbol
     * 
     *  Parameters:
     *  - symbol: Trading symbol
     *  - limit: Maximum number of reports
     * 
     *  Returns:
     *  - Array of report objects
     */
    async getReportsBySymbol(symbol, limit = 10) {
        try {
            const response = await apiClient.get(`/api/reports/symbol/${symbol}`, { 
                params: { limit } 
            });
            
            // Cache the latest report
            if (response.length > 0) {
                this.cache.set(symbol, response[0]);
            }
            
            return response;
        } catch (error) {
            Debugger.error(`Failed to fetch reports for ${symbol}: ${error.message}`);
            return [];
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │   GET LATEST REPORT BY SYMBOL       │
     *  └─────────────────────────────────────┘
     *  Fetches the most recent report for a symbol
     * 
     *  Parameters:
     *  - symbol: Trading symbol
     * 
     *  Returns:
     *  - Latest report object or null
     */
    async getLatestReportBySymbol(symbol) {
        try {
            // Check cache first
            if (this.cache.has(symbol)) {
                return this.cache.get(symbol);
            }
            
            const response = await apiClient.get(`/api/reports/symbol/${symbol}/latest`);
            
            // Update cache
            if (response) {
                this.cache.set(symbol, response);
            }
            
            return response;
        } catch (error) {
            Debugger.error(`Failed to fetch latest report for ${symbol}: ${error.message}`);
            return null;
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │         UPDATE REPORT               │
     *  └─────────────────────────────────────┘
     *  Updates an existing report
     * 
     *  Parameters:
     *  - reportId: Report ID
     *  - updates: Object with fields to update
     * 
     *  Returns:
     *  - Updated report object
     */
    async updateReport(reportId, updates) {
        try {
            const response = await apiClient.put(`/api/reports/${reportId}`, updates);
            
            // Update cache if needed
            if (response.symbol && this.cache.has(response.symbol)) {
                const cachedReport = this.cache.get(response.symbol);
                if (cachedReport.id === reportId) {
                    this.cache.set(response.symbol, response);
                }
            }
            
            Debugger.success(`Report ${reportId} updated`);
            return response;
        } catch (error) {
            Debugger.error(`Failed to update report: ${error.message}`);
            throw error;
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │         DELETE REPORT               │
     *  └─────────────────────────────────────┘
     *  Deletes a report
     * 
     *  Parameters:
     *  - reportId: Report ID
     * 
     *  Returns:
     *  - Success boolean
     */
    async deleteReport(reportId) {
        try {
            // Get report first to clear cache
            const report = await this.getReportById(reportId);
            
            await apiClient.delete(`/api/reports/${reportId}`);
            
            // Clear from cache
            if (report && report.symbol && this.cache.has(report.symbol)) {
                const cachedReport = this.cache.get(report.symbol);
                if (cachedReport.id === reportId) {
                    this.cache.delete(report.symbol);
                }
            }
            
            Debugger.success(`Report ${reportId} deleted`);
            return true;
        } catch (error) {
            Debugger.error(`Failed to delete report: ${error.message}`);
            return false;
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │    GET HIGH CONFIDENCE REPORTS      │
     *  └─────────────────────────────────────┘
     *  Fetches reports with high confidence
     * 
     *  Parameters:
     *  - minConfidence: Minimum confidence threshold
     *  - limit: Maximum number of reports
     * 
     *  Returns:
     *  - Array of high-confidence reports
     */
    async getHighConfidenceReports(minConfidence = 0.8, limit = 50) {
        try {
            const response = await apiClient.get(
                `/api/reports/high-confidence/${minConfidence}`,
                { params: { limit } }
            );
            return response;
        } catch (error) {
            Debugger.error(`Failed to fetch high confidence reports: ${error.message}`);
            return [];
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │         CLEAR CACHE                 │
     *  └─────────────────────────────────────┘
     *  Clears the report cache
     * 
     *  Parameters:
     *  - symbol: Optional - clear specific symbol only
     * 
     *  Returns:
     *  - None
     */
    clearCache(symbol = null) {
        if (symbol) {
            this.cache.delete(symbol);
        } else {
            this.cache.clear();
        }
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │      FORMAT REPORT FOR UI           │
     *  └─────────────────────────────────────┘
     *  Formats report data for UI display
     * 
     *  Parameters:
     *  - report: Report object
     * 
     *  Returns:
     *  - Formatted report object
     */
    formatReportForUI(report) {
        if (!report) return null;
        
        return {
            ...report,
            confidencePercent: Math.round(report.AIConfidence * 100),
            timeFetchedFormatted: new Date(report.timeFetched).toLocaleString(),
            actionClass: this.getActionClass(report.AIAction),
            confidenceClass: this.getConfidenceClass(report.AIConfidence)
        };
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │       GET ACTION CLASS              │
     *  └─────────────────────────────────────┘
     *  Gets CSS class for action type
     * 
     *  Parameters:
     *  - action: AI action (BUY, SELL, etc)
     * 
     *  Returns:
     *  - CSS class name
     */
    getActionClass(action) {
        const actionClasses = {
            'BUY': 'action-buy',
            'SELL': 'action-sell',
            'HOLD': 'action-hold',
            'WATCH': 'action-watch'
        };
        return actionClasses[action] || 'action-default';
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │     GET CONFIDENCE CLASS            │
     *  └─────────────────────────────────────┘
     *  Gets CSS class for confidence level
     * 
     *  Parameters:
     *  - confidence: Confidence value (0-1)
     * 
     *  Returns:
     *  - CSS class name
     */
    getConfidenceClass(confidence) {
        if (confidence >= 0.8) return 'confidence-high';
        if (confidence >= 0.6) return 'confidence-medium';
        return 'confidence-low';
    }

    /**
     * 
     *  ┌─────────────────────────────────────┐
     *  │      CHECK REPORTS AVAILABLE        │
     *  └─────────────────────────────────────┘
     *  Checks if reports are available for a symbol
     * 
     *  Parameters:
     *  - symbol: Trading symbol to check
     * 
     *  Returns:
     *  - Boolean indicating if reports exist
     */
    async checkReportsAvailable(symbol) {
        if (!symbol) return false;
        
        try {
            // Check cache first
            if (this.cache.has(symbol)) {
                return true;
            }
            
            // Fetch from API if not in cache
            const reports = await this.getReportsBySymbol(symbol, 1);
            return reports && reports.length > 0;
        } catch (error) {
            console.error(`Error checking reports for ${symbol}:`, error);
            return false;
        }
    }
}

// Export singleton instance
const reportsService = new ReportsService();
export { reportsService };
